File: compiler/TST/Structure.cs
public class Structure {

    private System.Object iOrigin;
    private System.Type iProto;
    private System.Reflection.ConstructorInfo iConstructor;

    static Structure(){
	System.Console.WriteLine("\nUSING Structure.cs!\n");
    }

    public Structure(System.Object o, System.Type p){
	iOrigin = o;
	iProto = p;
	iConstructor = null;
    }
  
    public static Structure AlloS(System.Object o, System.Type p) {
	return new Structure(o, p);
    }
  
    public static Structure ObjS(System.Object o){
	System.Type proto = o.GetType();
	System.Reflection.FieldInfo origin_field = proto.GetField("origin");
	System.Object origin = origin_field.GetValue(o);
	return new Structure(origin, proto);
    }

    public System.Object AlloSI(){
	System.Reflection.ConstructorInfo[] constructors = iProto.GetConstructors();
	if (iConstructor == null){
	  for (int i = 0; i < constructors.Length; i++) {
	    //System.Console.WriteLine("AlloSI: constructor: " + constructors[i].Name);
	    System.Reflection.ParameterInfo[] parameters = constructors[i].GetParameters();
	    if (parameters.Length == 1){
	      //System.Console.WriteLine("Has one parameter of type: " + parameters[0].ParameterType.FullName);
	      if (parameters[0].ParameterType.IsAssignableFrom(iOrigin.GetType())){
		iConstructor = constructors[i];
		break;
	      }
	    }
	  }
	}
#if false
	if (iConstructor==null){
	    System.Console.WriteLine("AlloSI:\tNo constructor with parameter type (origin)\n\t'" + iOrigin.GetType().FullName + "' found in class '" + iProto.FullName + "'!");
	}
#endif
	return iConstructor.Invoke(new System.Object[]{iOrigin});
    }
  
    public static bool eqS(Structure arg1, Structure arg2) {
	if (arg1==null) {
	    if (arg2==null)
		return true;
	    return false;
	}
	if (arg2==null)
	    return false;
	if (arg1.iProto != arg2.iProto)
	    return false;
	if (arg1.iOrigin != arg2.iOrigin)
	    return false;
	return true;
    }

    public static bool neS(Structure arg1, Structure arg2) {
	return !eqS(arg1, arg2);
    }

    public static bool leS(Structure arg1, Structure arg2) { 
	return (eqS(arg1, arg2) || ltS(arg1, arg2));
    }

    public static bool geS(Structure arg1, Structure arg2) { 
	return (eqS(arg1, arg2) || gtS(arg1, arg2));
    }

    public static bool gtS(Structure arg1, Structure arg2){
	return ltS(arg2, arg1);
    }
  
    public static bool ltS(Structure arg1, Structure arg2){
	System.Type proto1;
	System.Type proto2;
	if (arg1==null) return false;
	if (arg2==null) return false;
	proto1 = arg1.iProto;
	proto2 = arg2.iProto;
	if (proto1 == proto2) return false;
	  
	for (proto1 = proto1.BaseType;
	     proto1 != null;
	     proto1 = proto1.BaseType) {
	    if (proto1 == proto2) {
		if (false){
		  //return arg1.iOrigin == arg2.iOrigin;
		} else {
		    System.Object newObject = arg1.AlloSI();
		    System.Reflection.FieldInfo origin_field = proto1.GetField("origin");
		    return origin_field.GetValue(newObject) == arg2.iOrigin;
		}
	    }
	}
	return false; 
    }
}

File: compiler/TST/Structure.java
import java.lang.reflect.*;

public class Structure {

    public Object iOrigin;
    public Class iProto;
    private Constructor iConstructor;
  
    Structure(Object o, Class p){
	iOrigin = o;
	iProto = p;
	iConstructor = null;
    }

    public static Structure AlloS(Object o, Class p) {
	return new Structure(o, p);
    }

    public static Structure ObjS(Object o)
	throws Exception
    {
	Class proto = o.getClass();
	Field origin_field;
	Object origin;
	origin_field = proto.getDeclaredField("origin");
	origin = origin_field.get(o);
	return new Structure(origin, proto);
    }

    public Object AlloSI() throws Exception
    {
	Constructor[] constructors = iProto.getDeclaredConstructors();
	if (iConstructor == null){
	    search:
	    for (int i = 0; i < constructors.length; i++) {
		Class[] parameterTypes = constructors[i].getParameterTypes();
		if (parameterTypes.length == 1){
		    if (parameterTypes[0].isAssignableFrom(iOrigin.getClass())){
			iConstructor = constructors[i];
			break search;
		    }
		}
	    }
	}
	return iConstructor.newInstance(new Object[]{iOrigin});
    }

    public static boolean eqS(Structure arg1, Structure arg2)
    {
	if (arg1==null) {
	    if (arg2==null)
		return true;
	    return false;
	}
	if (arg2==null)
	    return false;
	if (arg1.iProto != arg2.iProto)
	    return false;
	if (arg1.iOrigin != arg2.iOrigin)
	    return false;
	return true;
    }

    public static boolean neS(Structure arg1, Structure arg2)
    {
	return !eqS(arg1, arg2);
    }

    public static boolean leS(Structure arg1, Structure arg2)
    { 
	return (eqS(arg1, arg2) || ltS(arg1, arg2));
    }

    public static boolean geS(Structure arg1, Structure arg2)
    { 
	return (eqS(arg1, arg2) || gtS(arg1, arg2));
    }

    public static boolean gtS(Structure arg1, Structure arg2)
    {
	return ltS(arg2, arg1);
    }

    public static boolean ltS(Structure arg1, Structure arg2)
    {
	Class proto1;
	Class proto2;
	Object newObject;

	if (arg1==null) return false;
	if (arg2==null) return false;
  
	proto1 = arg1.iProto;
	proto2 = arg2.iProto;
  
	if (proto1 == proto2)
	    return false;

	/* Prefix of proto1 is the first try */
  
	for (proto1 = proto1.getSuperclass();
	     proto1 != null; /* proto1 != Object## */
	     proto1 = proto1.getSuperclass()) {
	    if (proto1 == proto2) {
		/* There is some hope, now we need to check if origins are equal. */
		if (true /*FIXME: proto2->OriginOff == arg1->iProto->OriginOff*/){
		    /* The original prototypes have same origin offset 
		     * i.e. (same prefix level),
		     * so the result is (arg1->iOrigin == arg2->iOrigin) 
		     */
		    return arg1.iOrigin == arg2.iOrigin;
		}
	
		/* If proto1 and proto2 has different Origin Prefix
		 * we need to generate an object from arg1 and then
		 * test the resulting origin from the new object.
		 *   
		 * We need to generate a new item, as this is currently the only
		 * way we can get the origin.
		 * The problem is that there are several origins 
		 * (one per prefixlevel).
		 * So we generate an object corresponding to arg1 and find the 
		 * origin at the offset determined *by the prefix* (proto2). 
		 * This should be the same as origin of arg2.
		 */
	
		//newObject = arg1.AlloSI();
		//return true; // FIXME: (Object *)((long*)newObject)[proto2->OriginOff] == (arg2->iOrigin);
	    }
	}
    
	return false; 
    }

}
